File: CF/2245_D2.java
import java.io.*;
import java.util.*;

public class Main {
    public static void solve(int n, int m, int[][] edges, PrintWriter out) {
        ArrayList<int[]>[] graph = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }
        
        int[][] dis = new int[n][2];
        for (int i = 0; i < m; i++) {
            int o = edges[i][0];
            int u = edges[i][1];
            int v = edges[i][2];
            
            graph[u].add(new int[]{v, o});
            graph[v].add(new int[]{u, o});
            
            dis[u][o]++;
            dis[v][o]++;
        }

        ArrayDeque<Integer> q = new ArrayDeque<>();
        boolean[] vis = new boolean[n];
        for (int u = 0; u < n; u++) {
            if (dis[u][0] == 0 || dis[u][1] == 0) {
                vis[u] = true;
                q.offer(u);
            }
        }

        int[][] order = new int[n][2];
        int idx = 0;

        while (!q.isEmpty()) {
            int u = q.pollFirst();
            order[idx][0] = u;
            
            if (dis[u][0] == 0) {
                order[idx][1] = -1;
            } else {
                order[idx][1] = 1;
            }
            idx++;

            for (int i = 0; i < graph[u].size(); i++) {
                int[] edge = graph[u].get(i);
                int v = edge[0];
                int o = edge[1];
                
                if (--dis[v][o] == 0 && !vis[v]) {
                    vis[v] = true;
                    q.offer(v);
                }
            }
        }

        if (idx != n) {
            out.println("NO");
            return;
        }

        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            int revIdx = n - i - 1;
            res[order[revIdx][0]] = (i + 1) * order[revIdx][1];
        }
        
        out.println("YES");
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            sb.append(res[i]).append(i == n - 1 ? "" : " ");
        }
        out.println(sb.toString());
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int t = Integer.parseInt(st.nextToken());
        
        PrintWriter out = new PrintWriter(new BufferedOutputStream(System.out));
        
        while (t-- > 0) {
            while (!st.hasMoreTokens()) {
                st = new StringTokenizer(br.readLine());
            }
            int n = Integer.parseInt(st.nextToken());
            int m = Integer.parseInt(st.nextToken());
            
            int[][] edges = new int[m][3];
            for (int i = 0; i < m; i++) {
                while (!st.hasMoreTokens()) {
                    st = new StringTokenizer(br.readLine());
                }
                edges[i][0] = Integer.parseInt(st.nextToken()) - 1; // o
                edges[i][1] = Integer.parseInt(st.nextToken()) - 1; // u
                edges[i][2] = Integer.parseInt(st.nextToken()) - 1; // v
            }
            solve(n, m, edges, out);
        }
        out.flush();
    }
}

File: CF/2245_D2.py
import math
from collections import deque

t=int(input())  

def solve(n,m,edges):
    arr=[[0,0] for _ in range(n)]
    graph=[[] for _ in range(n)]
    
    for o,u,v in edges:
        arr[u][o]+=1
        arr[v][o]+=1
        graph[u].append((v,o))
        graph[v].append((u,o))
        
    q=deque()
    vis=[False]*n
    for u in range(n):
        if arr[u][0] == 0 or arr[u][1] == 0:
            vis[u]=True
            q.append(u)
    
    dirs=[]
    while q:
        u=q.popleft()
        if arr[u][0] == 0:
            dirs.append((u,-1))
        else:
            dirs.append((u,1)) 
        for v,o in graph[u]:
            arr[v][o]-=1
            if arr[v][o] == 0 and not vis[v]:
                vis[v]=True
                q.append(v)
    
    if len(dirs) < n:
        print("NO")
        return
    res=[0]*n
    dirs.reverse()
    for i in range(n):
        res[dirs[i][0]]=(i+1)*dirs[i][1]
    
    print("YES")
    print(*res)

for _ in range(t):
    n,m=map(int,input().split())
    edges=list(tuple(map(int,input().split())) for _ in range(m))
    edges=[(o-1,u-1,v-1) for o,u,v in edges]
    solve(n,m,edges)
